initial scaffolder docuentation (#720)
RoadieHQ/marketing-site: content/docs/details/scaffolder/index.md

---
title: Scaffolder
publishedDate: '2022-05-13T10:53:00.0Z'

> description: Initialize projects and trigger automation from within Backstage. You can prompt parameter inputs and combine scaffolder actions in oder to build a scaffolder template to run within Backstage.

Reviewer: ```suggestion
description: Initialize projects and trigger automation from within Backstage. You can prompt parameter inputs and combine scaffolder actions in order to build a scaffolder template to run within Backstage.
```

---

# Overview
Imagine you’re an engineer looking to create a new microservice. You want to get started as quickly as possible, with minimal boilerplate and red-tape to jump through. At the same time, engineering organizations benefit from having consistency in production, and often put gates in place to enforce it.

> Instead of creating blockers for engineering teams, Spotify use the scaffolder and quickly create new microservices, while helping to ensure that production remains mostly consistent.

Reviewer: ```suggestion
Instead of creating blockers for engineering teams, you can use the Roadie scaffolder and quickly create new microservices, while helping to ensure that production remains mostly consistent.
```

Engineers can choose a pre-defined software template, fill out a few form fields to provide values like the name of the GitHub repo that the new service will occupy, and click a button to run the template and create a new service.

# Components of a Template

> You can configure a Scaffolder _template_ to be triggered on demand by Backstage users. Your template contains one or more _steps_ which run sequentially during execution of a template. You may optionally specify All _steps_ during the template execution are running inside an ephemeral container that is destroyed after the execution completes.

Reviewer: ```suggestion
You can configure a Scaffolder _template_ to be triggered on demand by Roadie users. Your template contains one or more _steps_ which run sequentially during execution of a template. You may optionally specify All _steps_ during the template execution are running inside an ephemeral container that is destroyed after the execution completes.
```

> You can configure a Scaffolder _template_ to be triggered on demand by Backstage users. Your template contains one or more _steps_ which run sequentially during execution of a template. You may optionally specify All _steps_ during the template execution are running inside an ephemeral container that is destroyed after the execution completes.

Reviewer: ```suggestion
You can configure a Scaffolder _template_ to be triggered on demand by Backstage users. Your template contains one or more _steps_ which run sequentially during execution of a template. You may optionally specify all _steps_ while the template execution is running inside an ephemeral container that is destroyed after the execution completes.
```

A Scaffolder _template_ is a configurable process that will run one or more Scaffolder _steps_. The template will be run when a user visits the "Create Component" page in Backstage. `https://<tenant-name>.roadie.so/create`.

![create-a-new-component](./create-a-new-component.png)

> Templates are defined by a Backstage Entity YAML file with a _Template_ kind and imported into the Backstage catalog.  You can create multiple templates, each of which can perform a different set of steps. For example, you can have one _template_ that creates a React application, and another that creates a serverless app.

Reviewer: Casing of Template isn't consistent across the document. Should it be `Template` instead of _Template_ ?

> Templates are defined by a Backstage Entity YAML file with a _Template_ kind and imported into the Backstage catalog.  You can create multiple templates, each of which can perform a different set of steps. For example, you can have one _template_ that creates a React application, and another that creates a serverless app.

Reviewer: ```suggestion
Templates are defined by a Backstage Entity YAML file with a _Template_ kind and imported into the Backstage catalog. You can create multiple templates, each of which can perform a different set of steps. For example, you can have one _template_ that creates a React application, and another that creates a serverless app.
```

Here is an example of a very basic Scaffolder _template_ that prompts the user for a name, and then prints back the text "Hello, name!"

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: hello-world-template
  title: Hello World
  description: Says Hello to a specified name.
spec:
  owner: backstage/techdocs-core
  type: service

  parameters:
    - title: You are about to say hello to your first Backstage Template
      required:
        - name
      properties:
        name:
          title: Name
          type: string
          description: Name to say hello to
          ui:autofocus: true
          ui:options:
            rows: 5

  steps:
    - id: log-message
      name: Log Message
      action: debug:log
      input:

>         message: 'Hello, {{ parameters.name }}!'

Reviewer: ```suggestion
        message: 'Hello, ${{ parameters.name }}!'
```

According to https://backstage.io/docs/features/software-templates/migrating-from-v1beta2-to-v1beta3

```

# Header Section

The header section is required for every _template_ and contains information to configure the task and show details about the task on the "Create Component" page.

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: hello-world-template
  title: Hello World
  description: Says Hello to a specified name.
spec:
  owner: default/engineering
  type: service
```

## apiVersion
This is a required field and should be set to `scaffolder.backstage.io/v1beta3`

## kind
A Scaffolder template is also an Entity in Backstage. In order to configure this entity as a template you must set the kind to `Template`

## metadata
The metadata field contains some data that appears on the template card that appears on the "Create Component" page.

## spec
The spec field contains `owner` and `type`. Owner refers to the Backstage group or user that owns the Scaffolder task e.g. `default/engineering`. Type refers to the type of template. It can be set to anything and appears on the scaffolder template card in the "Create Component" page.

# Parameters

> The parameters is a list of parameter pages that can be prompted from the user when they run a template. Each array element contains the configuration for a single page of items to be filled by the user running the template. The parameter pages must contain `title`, `required` and `properties`.

Reviewer: ```suggestion
The parameters property is a list of parameter pages that can be prompted from the user when they run a template. Each array element contains the configuration for a single page of items to be filled by the user running the template. The parameter pages must contain `title`, `required` and `properties`.
```

Reviewer: What is a parameter 'page'?

Author: Hey. I think I was mistaken here. I think it's actually an array of parameters which may be an object or other types. Ill check this.

Parameters taken from the user earlier may be used in the action steps using the syntax `${{ parameters.name }}`.

## fetch:plain

Downloads content and places it in the workspace.

```yaml
steps
  - action: fetch:plain
    id: fetch-plain
    name: Fetch plain
    input:
      url: ./plain
```

Optionally, if you would prefer the data to be downloaded to a subdirectory in the workspace you may specify the 'targetPath' input option.

```yaml
steps
  - action: fetch:plain
    id: fetch-plain
    name: Fetch plain
    input:
      url: ./plain
      targetPath: fetched-data
```

## fetch:template

This downloads a directory containing templated files. It then renders all of the templates variables into the files and directory names and content, and places the result in the workspace.

```
steps
  - action: fetch:template
    id: fetch-template
    name: Fetch template
    input:
      url: ./template
      values:
        name: ${{ parameters.name }}
```

> The templated files themselves can contain refererces to the values in the following way `${{ values.name }}`. It uses TODO figure out templating language name and link it here.

Reviewer: ```suggestion
The templated files themselves can contain references to the values in the following way `${{ values.name }}`. It uses TODO figure out templating language name and link it here.
```

Optionally, if you would prefer the data to be downloaded to a subdirectory in the workspace you may specify the 'targetPath' input option.

```
steps
  - action: fetch:template
    id: fetch-template
    name: Fetch template
    input:
      url: ./template
      targetPath: fetched-data
      values:
        name: ${{ parameters.name }}
```

You can also choose to not template specific files downloaded by the task by using the `copyWithoutRender` option. It may use file paths or globs.

```
steps
  - action: fetch:template
    id: fetch-template
    name: Fetch template
    input:
      url: ./template
      copyWithoutRender:
        - README.md
        - src/**.ts
      values:
        name: ${{ parameters.name }}
```

If you would like to limit the templating to very specific files, you can optionally add the `.njk` extension to the files and use the `templateFileExtension` option.

```
steps
  - action: fetch:template
    id: fetch-template
    name: Fetch template
    input:
      url: ./template
      templateFileExtension: true
      values:
        name: ${{ parameters.name }}
```

## action:n

> TODO

Reviewer 2: Could direct them to the actions page in backstage?

---
title: Scaffolder

> publishedDate: '2022-05-13T10:53:00.0Z'

Reviewer 3: ```suggestion
publishedDate: '2022-05-16T10:53:00.0Z'
```

> description: Initialize projects and trigger automation from within Backstage. You can prompt parameter inputs and combine scaffolder actions in order to build a scaffolder template to run within Backstage.

Reviewer 3: ```suggestion
description: Initialize projects and trigger automation from within Backstage. You can prompt parameter inputs and combine scaffolder actions in order to build a scaffolder template.
```

Reviewer 3: To run within backstage was there twice

---

> # Overview

Reviewer 4: Best practice is to only have one `<h1 />` tag per page. The `title` from the YAML metadata is already rendered inside a `<h1 />` on this page. For content, your top level headings should be `<h2 />`.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#usage_notes

Author: will do.

Author: done

> Imagine you’re an engineer looking to create a new microservice. You want to get started as quickly as possible, with minimal boilerplate and red-tape to jump through. At the same time, engineering organizations benefit from having consistency in production, and often put gates in place to enforce it.

Reviewer 4: I'd probably be a bit more literal about what the scaffolder does here. This reads like a sales pitch rather than technical documentation.

Author: Sure. I copied it from the pitch :D

> Instead of creating blockers for engineering teams, you can use the Roadie scaffolder and quickly create new microservices, while helping to ensure that production remains mostly consistent.

Reviewer 4: Let's call it the Roadie Backstage scaffolder. We are not solely responsible for creating it.

Author: will do.

Engineers can choose a pre-defined software template, fill out a few form fields to provide values like the name of the GitHub repo that the new service will occupy, and click a button to run the template and create a new service.

# Components of a Template

> You can configure a Scaffolder _template_ to be triggered on demand by Roadie users. Your template contains one or more `steps` which run sequentially during execution of a template. You may optionally specify all `steps` while the template execution is running inside an ephemeral container that is destroyed after the execution completes.

Reviewer 3: > You may optionally specify all `steps`

What do you mean by this?

Author: I have no idea what I meant. I think maybe a merge fecked it up. Ill redo it.

> You can configure a Scaffolder _template_ to be triggered on demand by Roadie users. Your template contains one or more `steps` which run sequentially during execution of a template. You may optionally specify all `steps` while the template execution is running inside an ephemeral container that is destroyed after the execution completes.

Reviewer 4: I'd move this paragraph down a bit. You gotta explain what a template is before you say what the steps are.

> A Scaffolder template is a configurable process that will run one or more Scaffolder `steps`. The template will be run when a user visits the "Create Component" page in Backstage. `https://<tenant-name>.roadie.so/create`.

Reviewer 3: Duplication with the previous line?

spec:
  owner: backstage/techdocs-core
  type: service

  parameters:
    - title: You are about to say hello to your first Backstage Template
      required:
        - name
      properties:
        name:
          title: Name
          type: string
          description: Name to say hello to
          ui:autofocus: true
          ui:options:
            rows: 5

  steps:
    - id: log-message
      name: Log Message
      action: debug:log
      input:
        message: 'Hello, ${{ parameters.name }}!'
```

# Header Section

The header section is required for every `template` and contains information to configure the task and show details about the task on the "Create Component" page.

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: hello-world-template
  title: Hello World
  description: Says Hello to a specified name.
spec:
  owner: default/engineering
  type: service
```

## `apiVersion`
This is a required field and should be set to `scaffolder.backstage.io/v1beta3`

## `kind`
A Scaffolder template is also an Entity in Backstage. In order to configure this entity as a template you must set the kind to `Template`

> 

Reviewer 3: ```suggestion
```

## `metadata`
The metadata field contains some data that appears on the template card that appears on the "Create Component" page.

## `spec`
The spec field contains `owner` and `type`. Owner refers to the Backstage group or user that owns the Scaffolder task e.g. `default/engineering`. Type refers to the type of template. It can be set to anything and appears on the scaffolder template card in the "Create Component" page.

> # `parameters`

Reviewer 4: I find it confusing that "Header Section" (above) is a H1 title, whereas "parameters" is a code snippet in a H1. I would avoid using inline code in the header.

Author: yeah hmmm. I did this specifically for the side bar. Maybe we can do a design on this.

Author: Look at this page: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions

Reviewer 4: Fair enough!

```yaml
  parameters:
    properties:
      ownedEntity:
        type: string
        ui:field: EntityNamePicker
```

The respository picker can allow the user to select the name and location of a new respository. The picker restricts the target location of the repository to make it a little easier for the user to select a target location.

The following example, will only allow the user to enter a new repository name targetting the GitHub using the AcmeInc organization.
```yaml
  parameters:
    properties:
      repoUrl:
        type: string
        ui:field: RepoUrlPicker
        ui:options:
          allowedHosts:
            - github.com
          allowedOwners:
            - AcmeInc
```

The owner picker, allows the user to select a user / group in the Backstage catalog. e.g.
```yaml
  parameters:
    properties:
      owner:
        type: string
        ui:field: OwnerPicker
```

## `number`
You can allow the user to enter a number using the `number` type:

```yaml
  parameters:
    properties:
      size:
        type: number
```

## `object`

> The `object` allows the collection of more complex types of data from the user. It contains the `properies` option to add variables to the object as follows:

Reviewer 3: ```suggestion
The `object` allows the collection of more complex types of data from the user. It contains the `properties` option to add variables to the object as follows:
```

```

## `array`

You can prompt for an array of properties using the array option. The `items` option can be any type: `array`, `object`, `string` or `number` as you like.

```yaml
  parameters:
    properties:
      languages:
        type: array
        items:
          type: string
```

If you would like to prompt the user to add entity tags, you can use the `ui:field: EntityTagPicker` as shown below.

```yaml
  parameters:
    properties:
      entityTags:
        type: array
        ui:field: EntityTagsPicker
```

## Common Options

If you would like to default the value of a field you can use the `default` option:

```yaml
  parameters:
    properties:
      name:
        type: string
        default: "world!"
```

If you would like to prompt the users for a fixed list of options, you may use the `enum` option.

```yaml
  parameters:
    properties:
      size:
        type: number
        enum: [50, 100, 200]
```

You can display a more human description to a field value by using `title` and `description`

```yaml
  parameters:
    properties:
      name:
        type: string
        title: "Name"
        description: "Name to say hello to"
```

## Form Steps

> It might be jarring for your user to enter a lot of parameters one after another on the same page, especially if some of the propertiers require validation. As such Backstage have provided form steps.

Reviewer 3: ```suggestion
It might be jarring for your user to enter a lot of parameters one after another on the same page, especially if some of the properties require validation. As such Backstage have provided form steps.
```

You can make use of form steps using the following example.

```yaml
  parameters:
    - title: "Fill in the Name"
      properties:
        name:
          type: string
    - title: "Fill in the Age"
      properties:
        age:
          type: number
```

## More Reading

> You can read more about parameter configuration in the official backstage docs here https://backstage.io/docs/features/software-templates/writing-templates.

Reviewer 3: ```suggestion
You can read more about parameter configuration in the official backstage docs [here](https://backstage.io/docs/features/software-templates/writing-templates).
```

> You can read more about parameter configuration in the official backstage docs here https://backstage.io/docs/features/software-templates/writing-templates.

Reviewer 4: This functionality is provided to Backstage via another library right? Does it make sense to link to the docs of that library also?

Author: good idea.

Author: I looked into this a little. There is a sort of partial support for the uiSchema, and its behaviour only partially works. The linked page already links and references nuances about the uiSchema lib. So I think that is enough, what do you think?

> # `steps`

Reviewer 4: Same comment about not using inline code in titles.

Parameters taken from the user earlier may be used in the action steps using the syntax `${{ parameters.name }}`.

## `fetch:plain`

Downloads content and places it in the workspace.

```yaml
steps
  - action: fetch:plain
    id: fetch-plain
    name: Fetch plain
    input:
      url: ./plain
```

Optionally, if you would prefer the data to be downloaded to a subdirectory in the workspace you may specify the 'targetPath' input option.

```yaml
steps
  - action: fetch:plain
    id: fetch-plain
    name: Fetch plain
    input:
      url: ./plain
      targetPath: fetched-data
```

## `fetch:template`

This downloads a directory containing templated files. It then renders all of the templates variables into the files and directory names and content, and places the result in the workspace.

```
steps
  - action: fetch:template
    id: fetch-template
    name: Fetch template
    input:
      url: ./template
      values:
        name: ${{ parameters.name }}
```

> The templated files themselves can contain refererces to the values in the following way `${{ values.name }}`. It uses the nunjucks templating language name and link it here. More details can be found here: https://mozilla.github.io/nunjucks/

Reviewer 3: ```suggestion
The templated files themselves can contain refererces to the values in the following way `${{ values.name }}`. It uses the nunjucks templating language name and link it here. More details can be found [here]( https://mozilla.github.io/nunjucks/).
```

---

> title: Scaffolder

Reviewer 4: Should this title be a bit more descriptive? What am I going to learn about the scaffolder?

Author: How about now?

> title: Scaffolder - Tempalte Syntax

Reviewer 2: ```suggestion
title: Scaffolder - Creating Templates
```

publishedDate: '2022-05-16T10:53:00.0Z'
description: Initialize projects and trigger automation from within Backstage. You can prompt parameter inputs and combine scaffolder actions in order to build a scaffolder template.
---

## Overview

> The Roadie Backstage scaffolder is a feature that allows you to define software templates to create new software projects, update existing ones or simply perform repeated tasks in a consistent manner.

Reviewer 4: Does it actually allow you to update existing projects?

Author: Yes. you can create pull requests against existing projects. The action `publish:github:pull-request` is used by one of our customers to add resources to an existing terraform project based on a template and user input.

Author: Let me know if this is ok?

> Scaffolder templates are defined in yaml files and loaded into the Backstage catalog in the same way that other entities are loaded into Backstage. A template contains one or more `steps` which run sequentially during execution.

Reviewer 4: YAML should be upcased

```suggestion
Scaffolder templates are defined in YAML files and loaded into the Backstage catalog in the same way that other entities are loaded into Backstage. A template contains one or more `steps` which run sequentially during execution.
```
